### PR TITLE
Add support for conventions to ConfigurableFileCollection

### PR DESCRIPTION
**Pull request by abstratt**
Also allows convention mapping to work with CFCs.

See [spec](https://docs.google.com/document/d/1KHCCzNeNAIkpedH9fUcNUe_vgZAWHxCODzxVW_7NzLw/edit#heading=h.qutvq8vpnt95) for design discussions.

Fixes: #21752, #25509

### CODE REVIEW

File: platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy [hunk 1]
```diff
@@ -38,6 +38,21 @@ class FileCollectionIntegrationTest extends AbstractIntegrationSpec implements T
         type << ["Object", "Object[]", "Set", "LinkedHashSet", "List", "LinkedList", "Collection", "FileCollection"]
     }
 
+    def "can set elements as convention of file collection"() {
+        buildFile """
+            def files = objects.fileCollection().convention('b')
+            def elements = files.elements
+
+            def name = 'a'
+            files.configure { it.from(name) }
```
Comment (abstratt) [inline platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy:47]: @bamboo Any pointers on the incantation needed to set an implicit receiver for actions specified as Groovy closures?

Comment (bamboo) [reply platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy:47]: `files` needs to be decorated. That happens when it's created via the `ObjectFactory` or a properly configured `Instantiator`.

Comment (abstratt) [reply platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy:47]: Thanks, but not sure I follow. Isn't `files` above already created via `ObjectFactory`?

Comment (bamboo) [reply platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy:47]: Sorry, I didn't look the test close enough. Gradle doesn't decorate `Transformer` taking methods with `Closure` ones. That only happens for `Action` taking methods.

Comment (bamboo) [reply platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy:47]: FTR: https://github.com/gradle/gradle/blob/f0e72fb78af3019816aeff29018b2c70b4548b05/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/generator/AbstractClassGenerator.java#L906

**Comment by abstratt** [comment]
Unblocked by #27301 finally merging.

**Comment by abstratt** [comment]
@bot-gradle test RFN

File: platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy (hunk 1 above)
Comment (bamboo) [reply platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy:47]: Mystery solved, `ObjectFactory#fileCollection()` returns a non-decorated file collection: https://github.com/gradle/gradle/blob/9239189ef8d307b43b765f5b9637f509d8f5a436/platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/DefaultFileCollectionFactory.java#L112-L111

Comment (abstratt) [reply platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy:47]: Thanks - so what we have is ok, right? Or should we try to get the ability of setting the file collection as an implicit receiver some other way?

Comment (bamboo) [reply platforms/core-configuration/file-collections/src/integTest/groovy/org/gradle/api/file/FileCollectionIntegrationTest.groovy:47]: It seems this is the first method in `CFC` to require the capability, we should consider changing the way the collection is instantiated.

File: platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java [hunk 2]
```diff
@@ -75,10 +82,28 @@ public DefaultConfigurableFileCollection(@Nullable String displayName, PathToFil
...
+    }
+
+    @Override
+    public ConfigurableFileCollection configure(Action<FileCollectionConfigurer> action) {
+        action.execute(getActualValue());
+        return this;
+    }
+
+    @Override
+    public ConfigurableFileCollection configure(Closure<FileCollectionConfigurer> action) {
```
Comment (abstratt) [inline platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java:102]: @adammurdoch @bamboo Is this what you were suggesting today?

File: platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java [hunk 3]
```diff
@@ -211,23 +236,78 @@ public FileCollection plus(FileCollection collection) {
...
+    @Override
+    public ConfigurableFileCollection convention(Iterable<?> paths) {
+        assertMutable();
+        setConventionCollector(newValue(EMPTY_COLLECTOR, paths));
+        return this;
+    }
+
+    @Override
+    public ConfigurableFileCollection convention(Object... paths) {
+        setConventionCollector(newValue(EMPTY_COLLECTOR, paths));
```
Comment (abstratt) [inline platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java:251]: lacks call to `assertMutable()`

File: platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java [hunk 4]
```diff
@@ -211,23 +236,78 @@ public FileCollection plus(FileCollection collection) {
...
+
+    @Override
+    public SupportsConvention setToConvention() {
+        assertMutable();
+        value = valueState.setToConvention();
+        return this;
+    }
+
+    private void setConventionCollector(ValueCollector convention) {
+        assertMutable();
```
Comment (abstratt) [inline platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultConfigurableFileCollection.java:270]: Either `setConventionCollector()` does not need it, or `setExplicitCollector()` does

**Review by abstratt — COMMENTED** [review]
A couple of changes required

**Comment by abstratt** [comment]
@bot-gradle test RFN

**Comment by abstratt** [comment]
I will merge this as is, please keep the feedback coming, will address it in a follow-up PR

**Comment by abstratt** [comment]
@bot-gradle merge

**Comment by abstratt** [comment]
@bot-gradle merge

**Comment by abstratt** [comment]
Temporarily withdrawn by #28135

**Pull request merged**